drivers/rtc/rtc_ds3231.c

```diff
@@ -0,0 +1,868 @@
+		if (!(states[i] == 0 || states[i] == 1)) {
+			states[i] = -EINVAL;
+			err = -EINVAL;
+		}
+	}
+	return err;
+}
+
+static int alarm_set_callback(const struct device *dev, uint16_t id, rtc_alarm_callback cb, void *user_data)
+{
+	if (id < 0 || id >= ALARM_COUNT) {
+		return -EINVAL;
+	}
+
+	struct drv_data *data = dev->data;
+	data->alarms[id] = (struct alarm){
+		cb,
+		user_data
+	};
+
+	return 0;
+}
+
+static void check_alarms(const struct device *dev) {
+	struct drv_data *data = dev->data;
+
+	bool states[2];
+	get_alarm_states(dev, states);
+
+	for (int i = 0; i < ALARM_COUNT; i++) {
+		if (states[i]) {
+			if (data->alarms[i].cb) {
+				data->alarms[i].cb(dev, i, data->alarms[i].user_data);
+			}
+		}
+	}
+}
+static int init_alarms(struct drv_data *data) {
+	data->alarms[0] = (struct alarm){NULL, NULL};
+	data->alarms[1] = (struct alarm){NULL, NULL};
+	return 0;
+}
+#endif
+
+#ifdef CONFIG_RTC_UPDATE
+static int init_update(struct drv_data *data) {
+	data->update = (struct update){NULL, NULL};
+	return 0;
+}
+static int update_set_callback(const struct device *dev, rtc_update_callback cb, void *user_data)
+{
+	struct drv_data *data = dev->data;
+	data->update = (struct update){
+		cb,
+		user_data
+	};
+	return 0;
+}
+static void update_callback(const struct device *dev) {
+	struct drv_data *data = dev->data;
+	if (data->update.cb) {
+		data->update.cb(dev, data->update.user_data);
+	}
+}
+#endif /* CONFIG_RTC_UPDATE */
+
+#if defined(CONFIG_RTC_UPDATE) || defined(CONFIG_RTC_ALARM)
+static void isw_h(struct k_work *work)
+{
+	struct drv_data *data = CONTAINER_OF(work, struct drv_data, work);
+	const struct device *dev = data->dev;
+	
+#ifdef CONFIG_RTC_UPDATE
+	update_callback(dev);
+#endif /* CONFIG_RTC_UPDATE */
+
+#ifdef CONFIG_RTC_ALARM
+	check_alarms(dev);
+#endif /* CONFIG_RTC_ALARM */
+
+}
+static void isw_isr(const struct device *port, struct gpio_callback *cb, uint32_t pins)
+{
+	struct drv_data *data = CONTAINER_OF(cb, struct drv_data, isw_cb_data);
+
+	k_work_submit(&data->work);
+}
+static int init_isw(const struct drv_conf *config, struct drv_data *data) {
+	if (!gpio_is_ready_dt(&config->isw_gpios)) {
+		LOG_ERR("ISW GPIO pin is not ready.");
+		return -ENODEV;
+	}
+
+	k_work_init(&data->work, isw_h);
+
+	int err = gpio_pin_configure_dt(&(config->isw_gpios), GPIO_INPUT);
+	if (err != 0) {
+		LOG_ERR("Couldn't configure ISW GPIO pin.");
+		return err;
+	}
+	err = gpio_pin_interrupt_configure_dt(&(config->isw_gpios), GPIO_INT_EDGE_TO_ACTIVE);
+	if (err != 0) {
+		LOG_ERR("Couldn't configure ISW interrupt.");
+		return err;
+	}
+
+	gpio_init_callback(&data->isw_cb_data, isw_isr, BIT((config->isw_gpios).pin));
+	err = gpio_add_callback((config->isw_gpios).port, &data->isw_cb_data);
+	if (err != 0) {
+		LOG_ERR("Couldn't add ISW interrupt callback.");
+		return err;
+	}
+
+	return 0;
+}
+#endif /* defined(CONFIG_RTC_UPDATE) || defined(CONFIG_RTC_ALARM) */
+
+static const struct rtc_driver_api driver_api = {
```

Comment:
Device driver APIs need to be wrapped now

```suggestion
static DEVICE_API(rtc, driver_api) = {
```